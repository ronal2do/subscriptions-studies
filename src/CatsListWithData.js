// @flow
import React, { Component } from 'react'
import { graphql } from 'react-apollo'
import { CATS_SUBSCRIPTION, CATS_LIST_QUERY } from './Queries'
import ListItem from './ListItem'

type Props = {
  cats: CatsType,
}

type CatsType = {
  loading: boolean,
  error: ErrorType,
  allCats: Object,
  subscribeToMore: (config: Object) => void,
}

type ErrorType = {
  message: string,
}

class CatsList extends Component<Props> {
  componentDidMount () {
    this.subscribeCats()
  }

  render () {
    const { cats: { loading, error, allCats } } = this.props

    if (loading) {
      return <p>Loading ...</p>
    }
    if (error) {
      return <p>{error.message}</p>
    }

    return (
      <div className='list'>
        {allCats.edges.map(({ node }) => (
          <ListItem key={node.id} node={node} />
        ))}
      </div>
    )
  }

  subscribeCats = () => {
    this.props.cats.subscribeToMore({
      document: CATS_SUBSCRIPTION,
      updateQuery: (previous, { subscriptionData }) => {
        const updated = subscriptionData.data.newCat
        const node = [
          {
            node: updated,
            __typename: 'CatEdge'
          },
          ...previous.allCats.edges
        ]
        const result = {
          ...previous,
          allCats: {
            __typename: previous.allCats.__typename,
            edges: node
          }
        }
        return result
      }
    })
  }
}

export const CatsListWithData = graphql(CATS_LIST_QUERY, {
  name: 'cats'
})(CatsList)
