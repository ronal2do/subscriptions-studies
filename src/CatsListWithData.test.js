import React from 'react'
import ReactDOM from 'react-dom'
import { CatsListWithData } from './CatsListWithData'
import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

configure({ adapter: new Adapter() })

describe('<CatsListWithData  />', () => {
  const wrapper = shallow(<CatsListWithData />)

  it('should render CatsListWithData', () => {
    console.log(wrapper.debug())
  })

  it('calls componentDidMount', () => {
    jest.spyOn(CatsListWithData.prototype, 'componentDidMount')
    expect(CatsListWithData.prototype.componentDidMount.mock.calls.length).toBe(
      1
    )
  })

  it('Matches the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
