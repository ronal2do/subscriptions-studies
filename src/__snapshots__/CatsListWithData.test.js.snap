// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatsListWithData  /> Matches the snapshot 1`] = `
<Query
  displayName="Apollo(CatsList)"
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "allcats",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "allCats",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "edges",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "node",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "name",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "createdAt",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "avatarUrl",
                                  },
                                  "selectionSet": undefined,
                                },
                              ],
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 133,
        "start": 0,
      },
    }
  }
  skip={false}
  warnUnhandledError={true}
/>
`;
